src/runtime-rs/crates/hypervisor/src/device/topology.rs

```diff
@@ -0,0 +1,293 @@
+// Currently, CLH and Dragonball support device attachment solely on the root bus.
+const DEFAULT_PCIE_ROOT_BUS_ADDRESS: &str = "0000:00";
+pub const PCIE_ROOT_BUS_SLOTS_CAPACITY: u32 = 32;
+
+pub trait PCIeDevice: Send + Sync {
+    fn device_id(&self) -> &str;
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct PCIeEndpoint {
+    // device_id for device in device manager
+    pub device_id: String,
+    pub pci_path: PciPath,
+    // root_port for PCIe Device
+    pub root_port: Option<PCIeRootPort>,
+
+    // device_type is for device virtio-pci/PCI or PCIe
+    pub device_type: String,
+}
+
+impl PCIeDevice for PCIeEndpoint {
+    fn device_id(&self) -> &str {
+        self.device_id.as_str()
+    }
+}
+
+// reserved resource
+#[derive(Clone, Debug, Default)]
+pub struct ResourceReserved {
+    // This to work needs patches to QEMU
+    // The PCIE-PCI bridge can be hot-plugged only into pcie-root-port that has 'bus-reserve' property value to
+    // provide secondary bus for the hot-plugged bridge.
+    pub bus_reserve: String,
+
+    // reserve prefetched MMIO aperture, 64-bit
+    pub pref64_reserve: String,
+    // reserve prefetched MMIO aperture, 32-bit
+    pub pref32_reserve: String,
+    // reserve non-prefetched MMIO aperture, 32-bit *only*
+    pub memory_reserve: String,
+
+    // IO reservation
+    pub io_reserve: String,
+}
+
+// PCIe Root Port
+#[derive(Clone, Debug, Default)]
+pub struct PCIeRootPort {
+    // format: rp{n}, n>=0
+    pub id: String,
+
+    // default is pcie.0
+    pub bus: String,
+    // >=0, default is 0x00
+    pub address: String,
+
+    // (slot, chassis) pair is mandatory and must be unique for each pcie-root-port,
+    // chassis >=0, default is 0x00
+    pub chassis: u8,
+    // slot >=0, default is 0x00
+    pub slot: u8,
+
+    // multi_function is for PCIe Device passthrough
+    // true => "on", false => "off", default is off
+    pub multi_function: bool,
+
+    // reserved resource for some VMM, such as Qemu.
+    pub resource_reserved: ResourceReserved,
+
+    // romfile specifies the ROM file being used for this device.
+    pub romfile: String,
+}
+
+// PCIe Root Complex
+#[derive(Clone, Debug, Default)]
+pub struct PCIeRootComplex {
+    pub root_bus: String,
+    pub root_bus_address: String,
+    pub root_bus_devices: HashMap<String, PCIeEndpoint>,
+}
+
+#[derive(Debug, Default)]
+pub struct PCIeTopology {
+    pub hypervisor_name: String,
+    pub root_complex: PCIeRootComplex,
+
+    pub bridges: u32,
+    pub pcie_root_ports: u32,
+    pub hotplug_vfio_on_root_bus: bool,
+}
+
+impl PCIeTopology {
+    // As some special case doesn't support PCIe devices, there's no need to build a PCIe Topology.
+    pub fn new(config_info: Option<&TopologyConfigInfo>) -> Option<Self> {
+        // if config_info is None, it will return None.
+        let topo_config = config_info?;
+
+        let root_complex = PCIeRootComplex {
+            root_bus: DEFAULT_PCIE_ROOT_BUS.to_owned(),
```

Comment:
seems like all value in root_complex is default ones, can the default value cover all cases like the complex topology in qemu?

Reply:
more will be added when qemu is implemented in runtime-rs